File: NeetCode/Python/BinaryTrees/Easy/145.BinaryTreePostorder.py
# Given the root of a binary tree, return the postorder traversal of its nodes' values.

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def postorderTraversal(root):
    stack = [root]
    visit = [False]
    result = []

    while stack:
        current, visited = stack.pop(), visit.pop()

        if current:
            if visited:
                result.append(current.val)
            else:
                stack.append(current)
                visit.append(True)
                stack.append(current.right)
                visit.append(False)
                stack.append(current.left)
                visit.append(False)

    return result

File: swift/Neetcode/Trees/Easy/145.BinaryTreePostorder.swift
// Given the root of a binary tree, return the postorder traversal of its nodes' values.

// Time: O(n), Space: O(h)
func postorderTraversal(_ root: TreeNode?) -> [Int] {
    var result = [Int]()
    
    func dfs(_ root: TreeNode?) {
        guard let root = root else { return }

        dfs(root.left)
        dfs(root.right)
        result.append(root.val)
    }

    dfs(root)
    return result
}